Lib9c/Action/Buy5.cs

```diff
@@ -0,0 +1,224 @@
+            sellerAvatarAddress = plainValue["sellerAvatarAddress"].ToAddress();
+            productId = plainValue["productId"].ToGuid();
+        }
+
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            IActionContext ctx = context;
+            var states = ctx.PreviousStates;
+            if (ctx.Rehearsal)
+            {
+                states = states
+                    .SetState(buyerAvatarAddress, MarkChanged)
+                    .SetState(ctx.Signer, MarkChanged)
+                    .SetState(sellerAvatarAddress, MarkChanged)
+                    .MarkBalanceChanged(
+                        GoldCurrencyMock,
+                        ctx.Signer,
+                        sellerAgentAddress,
+                        GoldCurrencyState.Address);
+                return states.SetState(ShopState.Address, MarkChanged);
+            }
+            
+            var addressesHex = GetSignerAndOtherAddressesHex(context, buyerAvatarAddress, sellerAvatarAddress);
+
+            if (ctx.Signer.Equals(sellerAgentAddress))
+            {
+                throw new InvalidAddressException($"{addressesHex}Aborted as the signer is the seller.");
+            }
+
+            var sw = new Stopwatch();
+            sw.Start();
+            var started = DateTimeOffset.UtcNow;
+            Log.Verbose("{Addresses}Buy exec started", addressesHex);
+
+            if (!states.TryGetAvatarState(ctx.Signer, buyerAvatarAddress, out var buyerAvatarState))
+            {
+                throw new FailedLoadStateException($"{addressesHex}Aborted as the avatar state of the buyer was failed to load.");
+            }
+            sw.Stop();
+            Log.Verbose("{AddressesHex}Buy Get Buyer AgentAvatarStates: {Elapsed}", addressesHex, sw.Elapsed);
+            sw.Restart();
+
+            if (!buyerAvatarState.worldInformation.IsStageCleared(GameConfig.RequireClearedStageLevel.ActionsInShop))
+            {
+                buyerAvatarState.worldInformation.TryGetLastClearedStageId(out var current);
+                throw new NotEnoughClearedStageLevelException(
+                    addressesHex,
+                    GameConfig.RequireClearedStageLevel.ActionsInShop,
+                    current);
+            }
+
+            if (!states.TryGetState(ShopState.Address, out Bencodex.Types.Dictionary shopStateDict))
+            {
+                throw new FailedLoadStateException($"{addressesHex}Aborted as the shop state was failed to load.");
+            }
+
+            sw.Stop();
+            Log.Verbose("{AddressesHex}Buy Get ShopState: {Elapsed}", addressesHex, sw.Elapsed);
+            sw.Restart();
+
+            Log.Verbose(
+                "{AddressesHex}Execute Buy; buyer: {Buyer} seller: {Seller}",
+                addressesHex,
+                buyerAvatarAddress,
+                sellerAvatarAddress);
+            // 상점에서 구매할 아이템을 찾는다.
```

Comment:
Is it OK to keep comments written in Korean?

Reply:
i think this is not blocking issue.